File: source/ts/src/jyotish/natalBasis.ts
import { computeLagna } from './lagna';
import { computePlanetaryPositions } from './planets';
import { resolveMasaName, resolveNakshatraName } from '../i18n/resolver';
import { validateDate, validateLocation } from '../utils/validation';
import type { AyanamsaType, BirthChartOptions, Language } from '../types/options';
import type { GeoLocation } from '../types/location';
import type { LagnaInfo, PlanetaryPositions } from '../types/jyotish';

export interface NatalBasis {
  readonly birthDate: Date;
  readonly location: GeoLocation;
  readonly ayanamsaType: AyanamsaType;
  readonly lang: Language;
  readonly nodeType: 'mean' | 'true';
  readonly lagna: LagnaInfo;
  readonly positions: PlanetaryPositions;
}

export function computeNatalBasis(
  birthDate: Date,
  location: GeoLocation,
  options: BirthChartOptions = {},
): NatalBasis {
  validateDate(birthDate);
  validateLocation(location);

  const ayanamsaType: AyanamsaType = options.ayanamsa ?? 'lahiri';
  const lang: Language = options.language ?? 'en';
  const nodeType = options.nodeType ?? 'mean';

  return {
    birthDate,
    location,
    ayanamsaType,
    lang,
    nodeType,
    lagna: computeLagna(birthDate, location, ayanamsaType, lang),
    positions: computePlanetaryPositions(
      birthDate,
      ayanamsaType,
      (idx) => resolveNakshatraName(idx, lang),
      (idx) => resolveMasaName(idx, lang),
      nodeType,
    ),
  };
}

export function grahaList(basis: NatalBasis) {
  const p = basis.positions;
  return [
    { key: 'Sun' as const, pos: p.sun },
    { key: 'Moon' as const, pos: p.moon },
    { key: 'Mars' as const, pos: p.mars },
    { key: 'Mercury' as const, pos: p.mercury },
    { key: 'Jupiter' as const, pos: p.jupiter },
    { key: 'Venus' as const, pos: p.venus },
    { key: 'Saturn' as const, pos: p.saturn },
    { key: 'Rahu' as const, pos: p.rahu },
    { key: 'Ketu' as const, pos: p.ketu },
  ];
}

File: source/go/internal/jyotish/natalbasis.go
package jyotish

import (
	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/i18n"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
)

type BirthChartOptions struct {
	Ayanamsa    types.AyanamsaType
	Language    types.Language
	HouseSystem types.HouseSystem
	NodeType    NodeType
}

type NatalBasis struct {
	Ctx          *astronomy.EphemerisCtx
	BirthMs      int64
	Location     types.GeoLocation
	AyanamsaType types.AyanamsaType
	Lang         types.Language
	NodeType     NodeType
	Lagna        types.LagnaInfo
	Positions    types.PlanetaryPositions
}

func ComputeNatalBasis(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	options BirthChartOptions,
) (NatalBasis, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return NatalBasis{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return NatalBasis{}, err
	}

	ayanamsaType := resolveAyanamsa(options.Ayanamsa)
	lang := resolveLang(options.Language)
	nodeType := resolveNodeType(options.NodeType)

	lagna, err := ComputeLagna(ctx, birthMs, location, ayanamsaType, lang)
	if err != nil {
		return NatalBasis{}, err
	}
	positions, err := ComputePlanetaryPositions(ctx, birthMs, ayanamsaType,
		func(idx int) string { return i18n.ResolveNakshatraName(idx, lang) },
		func(idx int) string { return i18n.ResolveMasaName(idx, lang) },
		nodeType,
	)
	if err != nil {
		return NatalBasis{}, err
	}

	return NatalBasis{
		Ctx:          ctx,
		BirthMs:      birthMs,
		Location:     location,
		AyanamsaType: ayanamsaType,
		Lang:         lang,
		NodeType:     nodeType,
		Lagna:        lagna,
		Positions:    positions,
	}, nil
}

type GrahaEntry struct {
	Key types.Graha
	Pos *types.GrahaPosition
}

func GrahaList(basis *NatalBasis) []GrahaEntry {
	out := make([]GrahaEntry, 0, types.GrahaCount)
	for _, g := range types.AllGrahas {
		pos, ok := basis.Positions.Get(g)
		if !ok {
			panic("jyotish: AllGrahas contains a graha PlanetaryPositions.Get rejects")
		}
		out = append(out, GrahaEntry{Key: g, Pos: pos})
	}
	return out
}
